envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -31,40 +32,75 @@ class SnapshotUpdater(
         properties = properties
     )
 
-    fun start(changes: Flux<List<LocalityAwareServicesState>>): Flux<List<LocalityAwareServicesState>> {
-        // see GroupChangeWatcher
-        return Flux.combineLatest(
-            onGroupAdded,
-            changes,
-            BiFunction<Any, List<LocalityAwareServicesState>, List<LocalityAwareServicesState>> { _, states -> states }
-        )
-            .sample(properties.stateSampleDuration)
-            .publishOn(scheduler)
-            .doOnNext { states ->
-                versions.retainGroups(cache.groups())
-                updateSnapshots(states)
+    fun start(changes: Flux<List<LocalityAwareServicesState>>): Flux<UpdateResult> {
+        return Flux.merge(
+                services(changes),
+                groups()
+        ).doOnNext { result ->
+            val groups = if (result.action == Action.ALL_GROUPS) {
```

Comment:
Here we select if the update that was produced is from a service change and we need a rebuild or not.